packages/workbox-build/src/lib/write-sw.js

```diff
@@ -0,0 +1,85 @@
+const fse = require('fs-extra');
+const path = require('path');
+const template = require('lodash.template');
+
+const errors = require('./errors');
+const runtimeCachingConverter = require('./utils/runtime-caching-converter');
+
+module.exports = async (swDest, manifestEntries,
+                        pathToWorkboxSWFileRelativeToDest,
+                        globDirectory, options = {}) => {
+  try {
+    await fse.mkdirp(path.dirname(swDest));
+  } catch (error) {
+    throw new Error(`${errors['unable-to-make-sw-directory']}. ` +
+      `'${error.message}'`);
+  }
+
+  const templatePath = path.join(__dirname, 'templates', 'sw.js.tmpl');
+  let templateString;
+  try {
+    templateString = await fse.readFile(templatePath, 'utf8');
+  } catch (error) {
+    throw new Error(`${errors['read-sw-template-failure']}. ` +
+      `'${error.message}'`);
+  }
+  const workboxSWOptions = {};
+  if (options.cacheId) {
+    workboxSWOptions.cacheId = options.cacheId;
+  }
+  if (options.skipWaiting) {
+    workboxSWOptions.skipWaiting = true;
+  }
+  if (options.handleFetch === false) {
+    workboxSWOptions.handleFetch = false;
+  }
+  if (options.clientsClaim) {
+    workboxSWOptions.clientsClaim = true;
+  }
+  if (options.directoryIndex) {
+    workboxSWOptions.directoryIndex = options.directoryIndex;
+  }
+  if (options.ignoreUrlParametersMatching) {
+    // JSON.stringify can't output regexes so instead we'll
+    // inject it in the workboxSWOptionsString.
+    workboxSWOptions.ignoreUrlParametersMatching = [];
+  }
+
+  const runtimeCaching = runtimeCachingConverter(options.runtimeCaching);
+
+  let populatedTemplate;
+  try {
+    let workboxSWOptionsString = '';
+    if (Object.keys(workboxSWOptions).length > 0) {
+      workboxSWOptionsString = JSON.stringify(workboxSWOptions, null, 2);
+    }
+    if (options.ignoreUrlParametersMatching) {
+      workboxSWOptionsString = workboxSWOptionsString.replace(
+        '"ignoreUrlParametersMatching": []',
+        `"ignoreUrlParametersMatching": [` +
+            options.ignoreUrlParametersMatching.join(', ') + `]`
+      );
+    }
+    populatedTemplate = template(templateString)({
+      manifestEntries: manifestEntries,
+      workboxSWImportPath: pathToWorkboxSWFileRelativeToDest,
+      navigateFallback: options.navigateFallback,
+      navigateFallbackWhitelist: options.navigateFallbackWhitelist,
+      workboxSWOptionsString,
+      runtimeCaching,
+    }).trim() + '\n';
+  } catch (error) {
+    throw new Error(
+      `${errors['populating-sw-tmpl-failed']}. '${error.message}'`);
+  }
+
+  try {
+    await fse.writeFile(swDest, populatedTemplate);
```

Comment:
TODO: (Find a clean way to) Change the interface of this module so that it could just generate the SW string and return that, instead of always writing to disk.

This would unblock the Webpack usage, described at https://github.com/GoogleChrome/workbox/pull/808/files#r139604544

Reply:
Yes, if there is a way to import a `createSWTemplate` function or something that returns the value of `populatedTemplate` (given a `config` object), this would really help eliminate some code duplication in the webpack plugin.